### DISCUSSION TITLE
OIDC provider for multiple tenants

### DISCUSSION THREAD

**Discussion by johfer-bd** [authelia/authelia#7583] [Q&A] [answered]
Hi,

we have multiple customers -each with its own private web app instance- that I would like to authenticate with a single Authelia instance.

At this point, I'm only using the client_credentials grant type for machine-to-machine integration, i.e. no users, session, redirects etc. (human user auth is currently handled by AWS Cognito). Essentially I just want to create a token and validate/verify that token in the customer's web app instance. It should not be possible to use a token issued for customer1 in the web app of customer2 (obviously).

What is the proposed setup for this? Two options come to my mind:

* setting up multiple clients (one per customer), each one with its own audience (e.g. "customer1"). The web app needs to make sure to validate the audience against the allowed value(s) for the respective customer
* setting up multiple jwks keys (one per customer) which a matching client for each key. That way, each customer uses its own signature key and the web app somehow needs to check that the token has been signed by the expected key for that customer

I hope this makes sense, happy to elaborate if unclear. Thanks for your great work on this!

**Comment by james-d-elliott** [comment, marked as answer]
The audience claim is used for this purpose. Occasionally some apps use the scope too, however the audience is the correct method. For ID Tokens this is the client id, and for access tokens this is the granted audience.

There has not been a great deal of focus on the audience claim so in earlier implementations it was not strictly correct, and we're correcting it the deeper I understand the particular claim and how it's used within the landscape.

The audience should be validated against a URI when using the access token + introspection, and against a client identifier when checking if an ID Token is valid. It should be noted that the audience in an ID Token being the client ID per the spec is an indication of intent behind an ID token only being intended to be consumed by the client requesting it.

In addition I'd recommend using individual JWKs for each provider in this instance. It just makes it easier when you rotate them individually for any reason.

**Reply by johfer-bd** [reply]
Thank you, very useful!
